subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/messaging/SubscriptionMessagingInterface.kt

```diff
@@ -222,6 +223,43 @@ class SubscriptionMessagingInterface @Inject constructor(
         override val methods: List<String> = listOf("setSubscription")
     }
 
+    inner class SetAuthTokensMessage(
+        private val subscriptionsManager: SubscriptionsManager,
+        @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+        private val dispatcherProvider: DispatcherProvider,
+        private val pixelSender: SubscriptionPixelSender,
+        private val subscriptionsChecker: SubscriptionsChecker,
+    ) : JsMessageHandler {
+
+        override fun process(
+            jsMessage: JsMessage,
+            secret: String,
+            jsMessageCallback: JsMessageCallback?,
+        ) {
+            val (accessToken, refreshToken) = try {
+                with(jsMessage.params) { getString("accessToken") to getString("refreshToken") }
+            } catch (e: Exception) {
+                logcat { "Error parsing the tokens" }
+                return
+            }
+
+            appCoroutineScope.launch(dispatcherProvider.io()) {
+                try {
+                    subscriptionsManager.signInV2(accessToken, refreshToken)
+                    subscriptionsChecker.runChecker()
+                    pixelSender.reportRestoreUsingEmailSuccess()
+                    pixelSender.reportSubscriptionActivated()
+                } catch (e: Exception) {
+                    logcat { "Failed to set auth tokens" }
```

Comment:
is there a pixel we can send here to signal a problem and reason? or the methods we are using will already signal enough info to detect any unexpected failures?

Reply:
We do have some pixels that are fired inside SubscriptionManager to indicate a few specific errors that may happen. They don't necessarily cover all the possible failures. From the rollout perspective, the most important thing will be to monitor the `reportRestoreUsingEmailSuccess()` - if we notice a drop correlating with the rollout of this change, then we'll dig deeper, potentially adding more pixels.